src/libraries/System.Console/src/System/IO/ConsoleStream.cs

```diff
@@ -93,5 +93,37 @@ protected void ValidateWrite(byte[] buffer, int offset, int count)
 
             if (!_canWrite) throw Error.GetWriteNotSupported();
         }
+
+        private static void AccumulateNewLines(StringBuilder accumulator, ReadOnlySpan<char> buffer, Action<string> printer)
+        {
+            int lineStartIndex = 0;
+            for (int i = 0; i < buffer.Length; i++)
+            {
+                if (buffer[i] == '\n')
+                {
+                    ReadOnlySpan<char> sliceWithNl = buffer.Slice(lineStartIndex, i - lineStartIndex);
+                    if (accumulator.Length > 0)
+                    {
+                        // we found a new line, append content from accumulator if it's not empty
+                        accumulator.Append(sliceWithNl);
+                        printer(accumulator.ToString());
```

Comment:
There are lots of strings created. Would it be possible to pass ReadOnlySpan<char>s around instead?